From 98967e25e00bf005c1595c9acc7149f7c97ae0c4 Mon Sep 17 00:00:00 2001
From: Sriram R <quic_srirrama@quicinc.com>
Date: Sun, 29 Oct 2023 13:56:20 +0530
Subject: [PATCH 2/2] wifi: ath12k: Translate addr3 for AP properly

Currently only addr1/add2 are translated properly for AP mode.
With new hostapd changes in upstream which uses MLD addr in A3 as well,
ensure the addr3 is translated properly to the corresponding link bssid.

Signed-off-by: Sriram R <quic_srirrama@quicinc.com>
---
 drivers/net/wireless/ath/ath12k/mac.c | 4 ++++
 1 file changed, 4 insertions(+)

diff --git a/drivers/net/wireless/ath/ath12k/mac.c b/drivers/net/wireless/ath/ath12k/mac.c
index bd87677..e29ab63 100644
--- a/drivers/net/wireless/ath/ath12k/mac.c
+++ b/drivers/net/wireless/ath/ath12k/mac.c
@@ -9236,6 +9236,8 @@ static u8 ath12k_mac_get_tx_link(struct ieee80211_sta *sta, struct ieee80211_vif
 			ether_addr_copy(hdr->addr2, bss_conf->addr);
 			if (vif->type == NL80211_IFTYPE_STATION && bss_conf->bssid)
 				ether_addr_copy(hdr->addr3, bss_conf->bssid);
+			else if (vif->type == NL80211_IFTYPE_AP)
+				ether_addr_copy(hdr->addr3, bss_conf->addr);
 		} else if (bss_conf) {
 			/* In certain cases where a ML sta associated and added subset of links
 			 * on which the ML AP is active, but now sends some frame (ex. Probe request)
@@ -9254,6 +9256,8 @@ static u8 ath12k_mac_get_tx_link(struct ieee80211_sta *sta, struct ieee80211_vif
 			ether_addr_copy(hdr->addr2, bss_conf->addr);
 			if (vif->type == NL80211_IFTYPE_STATION && bss_conf->bssid)
 				ether_addr_copy(hdr->addr3, bss_conf->bssid);
+			else if (vif->type == NL80211_IFTYPE_AP)
+				ether_addr_copy(hdr->addr3, bss_conf->addr);
 
 
 			ath12k_dbg(NULL, ATH12K_DBG_MAC,
-- 
2.17.1

